### ISSUE TITLE
Provide a list of recommendations to reduce memory allocations.

### ISSUE THREAD

**Issue by dv00d00** [AngleSharp/AngleSharp#1151]
### Description

I am migrating my asp.net core app to .net 8 and in the spirit of this migration I've stumbled across the fact that parsing a 250 kb with my standard approach consumes around 5.3 MB of RAM.

I've started experimenting with the benchmark and discovered several potential improvements to my code.

To provide some context: 

```csharp

using AngleSharp;
using AngleSharp.Dom;
using AngleSharp.Html.Parser;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Environments;
using BenchmarkDotNet.Jobs;

namespace ConsoleApp7;

[Config(typeof(Config))]
[MemoryDiagnoser]
public class AngleSharpBenchmarks
{
    private class Config : ManualConfig
    {
        public Config()
        {
            AddJob(Job.ShortRun.WithRuntime(CoreRuntime.Core80));
        }
    }
    
    private static readonly HtmlParserOptions Options = new HtmlParserOptions()
    {
        IsEmbedded = false,
        IsScripting = false,
        IsStrictMode = false,
        IsNotSupportingFrames = true,
        IsKeepingSourceReferences = false,
        IsPreservingAttributeNames = false,
        IsSupportingProcessingInstructions = false,
        IsNotConsumingCharacterReferences = true,
        IsAcceptingCustomElementsEverywhere = true,
    };
    private static readonly BrowsingContext Context = new BrowsingContext();
    private static readonly HtmlParser parser = new HtmlParser(Options, Context);
    
    [Benchmark(Description = "Custom options + no dispose + unwrap linq + cached parser")]
    public async Task<string?> Allocate_Cached()
    {
        await using var file = File.OpenRead("home.html");
        var document = await parser.ParseDocumentAsync(file);
        
        string? found = null;
        foreach (var scriptElement in document.Scripts)
        {
            var data = scriptElement.ChildNodes.OfType<IText>().FirstOrDefault()?.Data;
            var contains = data?.Contains("var dimensionsArray = [];");
            if (contains == true)
            {
                found = data;
                break;
            }
        }

        return found;
    }
    
    [Benchmark(Description = "parse stream + no dispose + unwrap linq")]
    public async Task<string?> Allocate_Realistic()
    {
        await using var file = File.OpenRead("home.html");
        var parser = new HtmlParser();
        var document = await parser.ParseDocumentAsync(file);
        string? found = null;
        foreach (var scriptElement in document.Scripts)
        {
            var data = scriptElement.ChildNodes.OfType<IText>().FirstOrDefault()?.Data;
            var contains = data?.Contains("var dimensionsArray = [];");
            if (contains == true)
            {
                found = data;
                break;
            }
        }

        return found;
    }

    [Benchmark(Description = "parse stream + no dispose")]
    public async Task<string?> Allocate()
    {
        await using var file = File.OpenRead("home.html");
        var parser = new HtmlParser();
        var document = await parser.ParseDocumentAsync(file);
        var interestingScript = document.Scripts.FirstOrDefault(it => it.GetDescendants().OfType<IText>().FirstOrDefault()?.Data.Contains("var dimensionsArray = [];") == true);
        return interestingScript?.Text;
    }
    
    [Benchmark(Description = "parse stream + dispose doc")]
    public async Task<string?> Allocate4()
    {
        await using var file = File.OpenRead("home.html");
        var parser = new HtmlParser();
        using var document = await parser.ParseDocumentAsync(file);
        var interestingScript = document.Scripts.FirstOrDefault(it => it.Text.Contains("var dimensionsArray = [];"));
        return interestingScript?.Text;
    }
    
    [Benchmark(Description = "parse string", Baseline = true)]
    public async Task<string?> Allocate5()
    {
        var file = await File.ReadAllTextAsync("home.html");
        var parser = new HtmlParser();
        var document = parser.ParseDocument(file);
        var interestingScript = document.Scripts.FirstOrDefault(it => it.Text.Contains("var dimensionsArray = [];"));
        return interestingScript?.Text;
    }
    
    [Benchmark(Description = "parse string + using on doc")]
    public async Task<string?> Allocate6()
    {
        var file = await File.ReadAllTextAsync("home.html");
        var parser = new HtmlParser();
        using var document = parser.ParseDocument(file);
        var interestingScript = document.Scripts.FirstOrDefault(it => it.Text.Contains("var dimensionsArray = [];"));
        return interestingScript?.Text;
    }
}
```

and results:

```
| Method                                            | Mean     | Error     | StdDev    | Ratio | RatioSD | Gen0     | Gen1     | Gen2     | Allocated | Alloc Ratio |
|-------------------------------------------------- |---------:|----------:|----------:|------:|--------:|---------:|---------:|---------:|----------:|------------:|
| 'Custom options + no using + unwrap linq'         | 5.896 ms | 2.4853 ms | 0.1362 ms |  0.68 |    0.03 | 476.5625 | 265.6250 | 125.0000 |   3.65 MB |        0.76 |
| 'Custom options + no using'                       | 5.890 ms | 0.1473 ms | 0.0081 ms |  0.68 |    0.01 | 476.5625 | 320.3125 | 156.2500 |   3.76 MB |        0.78 |
| 'Default options'                                 | 5.947 ms | 0.1298 ms | 0.0071 ms |  0.69 |    0.01 | 492.1875 | 328.1250 | 328.1250 |   4.49 MB |        0.94 |
| 'Cached options + cached context'                 | 5.973 ms | 0.7051 ms | 0.0386 ms |  0.69 |    0.01 | 492.1875 | 328.1250 | 328.1250 |   4.43 MB |        0.92 |
| 'Cached options + cached context + cached parser' | 5.957 ms | 1.0465 ms | 0.0574 ms |  0.69 |    0.01 | 492.1875 | 328.1250 | 328.1250 |   4.43 MB |        0.92 |
| string                                            | 8.666 ms | 2.2499 ms | 0.1233 ms |  1.00 |    0.00 | 812.5000 | 796.8750 | 421.8750 |    4.8 MB |        1.00 |
| 'string + using on doc'                           | 7.019 ms | 1.5742 ms | 0.0863 ms |  0.81 |    0.00 | 906.2500 | 820.3125 | 531.2500 |   5.28 MB |        1.10 |
```
Disposing a documents results in an 10% of memory consumption.
Accessing text content might be pricey 
Caching parser + settings yields a minor memory consumption win, but it is unclear if I can cache it. Could it be accumulating circular references, idk.

Would appreciate if someone can come up with answer to the following questions:

1. Is it safe to not dispose `IDocument` instances?
2. Can I reuse parser instances during the lifetime of an app? Is there a sample on how to pool them maybe?
3. Is there a cheap way or maybe an overload which will allow to get text content of a script without extra allocations?
4. What options on `HtmlParserOptions` and `BrowsingContext` can/should I use to reduce memory consumption if I am only interested in parsing html (no scripting and css) without preserving context.

**Comment by FlorianRappl** [comment]
> Is it safe to not dispose IDocument instances?

It should be pretty safe, but disposing it should be a bit more reliable. It would be great if somebody does some tests on this.

> Can I reuse parser instances during the lifetime of an app? Is there a sample on how to pool them maybe?

The parser yes - the underlying DOM builder no (but I guess you refer to the parser itself, which you can definitely re-use). You can even use a single parser for everything - as the parser is just an abstraction and spawns new DOM builders there would be no side-effect. However, it is not thread-safe and it is bound to a given browsing context and options. So, if you want to change either one of them then don't reuse it.

> Is there a cheap way or maybe an overload which will allow to get text content of a script without extra allocations?

No - the parser always constructs the full model and thus you will always get the allocations for everything in the DOM. Presumably, we could maybe add something to not allocate / immediately throw elements away, but the use-case for this is little and the effort would be high.

> What options on HtmlParserOptions and BrowsingContext can/should I use to reduce memory consumption if I am only interested in parsing html (no scripting and css) without preserving context.

The standard options should be as good as it can be. For instance, switching `IsKeepingSourceReferences` to `true` would certainly have a bad impact, so don't do that unless you need to keep the original data around (with additional pointers from essentially every element).

Likewise, the `Configuration.Default` should be quite good, but if you want to be even better then throw out everything related to CSS selectors / parsing. Nevertheless, here I also don't see much saving potential. In the end these are tiny allocations and the CSS selector parsing would only be kicked off if you run something like `myElement.QuerySelector(...)`.

**Comment by dv00d00** [comment]
I've been using AngleSharp for a several years already, and was not disposing parsed documents, just started doing it recently for `reasons`. I've deployed a version to production env without `Dispose`. Last time it was causing issues it was due to GC not removing parsed docs due to circular references probably, (#45). To be fair I was not noticing any memory related issues with it, but will keep an eye on it. Common symptom is memory working set grows indefinitely?

Due to parser reuse. If, lets say, I was always using `HtmlParser` with default config I can potentially declare a factory with a `ObjectPool<HtmlParser>` implementation and return rented parsers there, guarantying no multithreaded access to it I should be fine? It won't accumulate state over time somehow? (sorry for maybe asking same question twice but just to be sure)
